Add QED particle creaction routines

lucafedeli88 opened the pull request:
This pull request implements QED particle creation routines (pair generation via the Breit-Wheeler process and photon emission). 

#### Main changes:
- The `QED` folder in `Source` has been moved to `Source/Particles/ElementaryProcess/QEDInternals` 

- `Source/Particles/ElementaryProcess/QEDPhotonEmission.H` provides functors to be used with the new particle creation routines to implement photon emission.

- `Source/Particles/ElementaryProcess/QEDPairGeneration.H` provides functors to be used with the new particle creation routines to implement pair generation.

- In `MultiParticleContainer.H` a new version of `MFItInfoCheckTiling` has been implemented, in order to take into account multiple destination species.

- In  `MultiParticleContainer.cpp`, the  `mapSpeciesProduct ` function has been modified to pair source and product species for the QED processes

- In `MultiParticleContainer.cpp` the two functions `doQedBreitWheeler` and `doQedQuantumSync` have been implemented. It is worth to remark that `doQedQuantumSync` calls `cleanLowEnergyPhotons`. The purpose of this function is to remove right after creation low energy photons which cannot generate pairs. This is achieved by setting to -1 the id of the low energy photons. 

- In `FilterCopyTransform.H ` some functions have been modified to better suit the case of two destination species.

- QED tests have been updated. **Warning** : QED tests are incomplete. For Breit Wheeler process only optical depth initialization and evolution are checked. For photon emission only optical depth initialization is checked.

- In `PhysicalParticleContainer` an `enum  PhysicalParticleType` has been added. The possible values are `photon`, `electron`, `positron`, `other`. Constructors of `PhysicalParticleContainer` and `PhotonParticleContainer` set a private `PhysicalParticleType` variable. This variable is used in methods `AmIAPhoton`, `AmIAPositron` and `AmIAnElectron`. This strategy has been adopted to fix an issue related to virtual function calls in constructors found by `lgtm`.

ax3l commented on Source/Particles/MultiParticleContainer.H:
```diff
@@ -253,15 +273,33 @@ protected:
+
+        MFItInfoCheckTiling(pc_src, pc_dsts...);
+
+    if (pc_src.do_tiling && amrex::Gpu::notInLaunchRegion()) {
```
alignment slightly off :)

lucafedeli88 replied:
ah, you are right!

ax3l commented on Source/Particles/MultiParticleContainer.H:
```diff
@@ -253,15 +273,33 @@ protected:
     std::shared_ptr<QuantumSynchrotronEngine> m_shr_p_qs_engine;
     //_______________________________
 
+
```
```suggestion
```

ax3l commented on Source/Particles/MultiParticleContainer.cpp:
```diff
@@ -1019,4 +1044,163 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
+
+                setNewParticleIDs(dst_tile, np_dst, num_added);
+
+                const auto energy_threshold = static_cast<amrex::ParticleReal>(
```
```suggestion
                constexpr auto energy_threshold = amrex::ParticleReal(
```

ax3l commented on Source/Particles/PhysicalParticleContainer.H:
```diff
@@ -332,12 +347,17 @@ protected:
     bool do_classical_radiation_reaction = false;
 
+    PhysicalParticleType m_my_type = PhysicalParticleType::other;
+    bool AmIAPhoton () {return (m_my_type == PhysicalParticleType::photon);};
```
```suggestion
    bool AmIAPhoton () override {return (m_my_type == PhysicalParticleType::photon);};
```

ax3l commented on Source/Particles/PhysicalParticleContainer.H:
```diff
@@ -332,12 +347,17 @@ protected:
 
+    PhysicalParticleType m_my_type = PhysicalParticleType::other;
+    bool AmIAPhoton () {return (m_my_type == PhysicalParticleType::photon);};
+    bool AmIAnElectron () {return (m_my_type == PhysicalParticleType::electron);};
```
```suggestion
    bool AmIAnElectron () override {return (m_my_type == PhysicalParticleType::electron);};
```

ax3l commented on Source/Particles/PhysicalParticleContainer.H:
```diff
@@ -332,12 +347,17 @@ protected:
+    PhysicalParticleType m_my_type = PhysicalParticleType::other;
+    bool AmIAPhoton () {return (m_my_type == PhysicalParticleType::photon);};
+    bool AmIAnElectron () {return (m_my_type == PhysicalParticleType::electron);};
+    bool AmIAPositron () {return (m_my_type == PhysicalParticleType::positron);};
```
```suggestion
    bool AmIAPositron () override {return (m_my_type == PhysicalParticleType::positron);};
```

ax3l commented on Source/Particles/PhysicalParticleContainer.H:
```diff
@@ -332,12 +347,17 @@ protected:
     //This function return true if the PhysicalParticleContainer contains electrons
     //or positrons, false otherwise
-    virtual bool AmIALepton ();
+
```
```suggestion
```

ax3l commented on Source/Particles/PhysicalParticleContainer.H:
```diff
@@ -332,12 +347,17 @@ protected:
+    bool AmIAPhoton () {return (m_my_type == PhysicalParticleType::photon);};
+    bool AmIAnElectron () {return (m_my_type == PhysicalParticleType::electron);};
+    bool AmIAPositron () {return (m_my_type == PhysicalParticleType::positron);};
+
```
```suggestion
```

MaxThevenet commented:
Just a small note: there could be conflicts to solve with PR https://github.com/ECP-WarpX/WarpX/pull/738.

lucafedeli88 commented on Source/Particles/ElementaryProcess/QEDPairGeneration.H:
```diff
@@ -0,0 +1,199 @@
+    void operator() (DstData& dst1, DstData& dst2, const SrcData& src,
+        int i_src, int i_dst1, int i_dst2) const noexcept
+    {
+
```
you are right... I should do that

lucafedeli88 commented on Source/Particles/ElementaryProcess/QEDPhotonEmission.H:
```diff
@@ -0,0 +1,221 @@
+    }
+
+private:
+    int m_opt_depth_runtime_comp;
```
Sorry, I forgot to add this!

lucafedeli88 commented on Source/Particles/ElementaryProcess/QEDPhotonEmission.H:
```diff
@@ -0,0 +1,221 @@
+    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    void operator() (DstData& dst, SrcData& src, int i_src, int i_dst) const noexcept
+    {
+
```
Thanks for the advice!

lucafedeli88 commented:
>     * For the species type (electron, positron, etc.), I think it would make sense to have these as input parameters instead of guessing from mass and charge. I can make a PR for that.

I totally agree! Thanks for https://github.com/ECP-WarpX/WarpX/pull/746 , as soon as it is merged I will make the necessary modifications. 

>     * As discussed before, I think it is confusing to use `tau` for two different quantities, I'll make a PR to use two different names.

I agree. Thanks for https://github.com/ECP-WarpX/WarpX/pull/742 ! As above, as soon as it is merged I will adapt this PR.
 
>     * I did not get into the details, but it seems to me that the tau initialization (`enable_opt_depth_for_targets` etc.) could be massively simplified by the `InitializationPolicy` that @atmyers introduced in PR #598. To my understanding, you can provide an initialization policy to particle quantities, and SmartCopy will apply this policy when doing the initialization, if the attribute exists in `dst`. @atmyers is that correct? I think things are made more complicated here because we use the name `tau` for two different things, which should be fixed, as mentioned above. @lucafedeli88 do you think you could use the `InitializationPolicy`?

That is correct. `tau` initialization is just a random number extraction from an exponential distribution. Actually, the same functor could be used for both electrons/positrons and photons. However, I prefer to keep them separate in case of future changes... As soon as  https://github.com/ECP-WarpX/WarpX/pull/742 is merged I will address this issue. As you mentioned this should make some methods much simpler...

lucafedeli88 commented on Source/Particles/MultiParticleContainer.cpp:
```diff
@@ -1019,4 +1044,163 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
+
+        auto Filter    = phys_pc_ptr->getPairGenerationFilterFunc();
+        auto CopyEle      = copy_factory_ele.getSmartCopy();
+        auto CopyPos      = copy_factory_pos.getSmartCopy();
```
we can try to make them `const` and see what happens...

lucafedeli88 replied:
apparently, some of them can be `const`...

lucafedeli88 commented on Source/Particles/ElementaryProcess/QEDPhotonEmission.H:
```diff
@@ -0,0 +1,221 @@
+        int opt_depth_runtime_comp,
+        QuantumSynchrotronGeneratePhotonAndUpdateMomentum emission_functor
+        ):
+        m_opt_depth_functor{opt_depth_functor},
```
Actually optical depth initialization functors don't have member variables, so I guess that I can simply pass-by-copy here...

lucafedeli88 commented:
Thanks a lot for your reviews! I should have addressed all the points that you've raised. However, I have to wait for a couple of other PRs to be merged before performing some of the changes that you have suggested...

MaxThevenet commented:
Is this still WIP?

MaxThevenet commented on Source/Particles/MultiParticleContainer.H:
```diff
@@ -22,9 +22,9 @@
-#   include "QED/QuantumSyncEngineWrapper.H"
+#   include "QedChiFunctions.H"
+#   include "BreitWheelerEngineWrapper.H"
+#   include "QuantumSyncEngineWrapper.H"
```
```suggestion
#   include "Particles/ElementaryProcess/QEDInternals/QedChiFunctions.H"
#   include "Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.H"
#   include "Particles/ElementaryProcess/QEDInternals/QuantumSyncEngineWrapper.H"
```
Now we add the paths to include statements.

lucafedeli88 replied:
Ok! I will comply with this policy in the future... I could not commit directly your suggestion because only `QedChiFunctions.H` is in `QEDInternals`, but I have implemented it.

MaxThevenet commented on Source/Particles/MultiParticleContainer.cpp:
```diff
@@ -948,4 +973,136 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
+
+void MultiParticleContainer::doQedBreitWheeler()
+{
+    WARPX_PROFILE("MPC::doQedEvents::doQedBreitWheeler");
```
```suggestion
    WARPX_PROFILE("MPC::doQedBreitWheeler");
```

MaxThevenet commented on Source/Particles/MultiParticleContainer.cpp:
```diff
@@ -948,4 +973,136 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
+#endif
+            for (MFIter mfi = pc_source->MakeMFIter(lev, info); mfi.isValid(); ++mfi)
+            {
+                auto& src_tile = pc_source ->ParticlesAt(lev, mfi);
```
```suggestion
                const auto& src_tile = pc_source->ParticlesAt(lev, mfi);
```
It may be worth trying putting a bunch of `const` here and there.

lucafedeli88 replied:
Unfortunately it seems that `src_tile` and `dst_tile` cannot be `const` due to the fact that they are passed to `filterCopyTransformParticles`. However, I've added a bunch of `const` as requested!

MaxThevenet commented on Source/Particles/MultiParticleContainer.cpp:
```diff
@@ -948,4 +973,136 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
+                setNewParticleIDs(dst_tile, np_dst, num_added);
+
+                constexpr auto energy_threshold = amrex::ParticleReal(
+                     2.0 * PhysConst::m_e * PhysConst::c * PhysConst::c);
```
```suggestion
                constexpr auto energy_threshold = static_cast<amrex::ParticleReal>(
                     2.0 * PhysConst::m_e * PhysConst::c * PhysConst::c);
```
?

lucafedeli88 replied:
ah sure!

MaxThevenet commented on Source/Particles/PhotonParticleContainer.cpp:
```diff
@@ -29,13 +29,22 @@ PhotonParticleContainer::PhotonParticleContainer (AmrCore* amr_core, int ispecie
     : PhysicalParticleContainer(amr_core, ispecies, name)
 {
 
+    m_my_type = PhysicalParticleType::photon;
```
Could use `WarpXParticleContainer::physical_species` directly? It should be meant for that.

lucafedeli88 replied:
ah, yes! sure...

lucafedeli88 replied:
I've also replaced all the `AmISomething` functions with a unique: 
```
    template<PhysicalSpecies PhysSpec>
    bool AmIA () const noexcept {return (physical_species == PhysSpec);}
```
in `WarpXParticleContainer.H`

MaxThevenet replied:
That look nice.

MaxThevenet commented on Source/Particles/PhysicalParticleContainer.H:
```diff
@@ -21,6 +21,9 @@
     #include <QedChiFunctions.H>
+
+    #include <QEDPairGeneration.H>
+    #include <QEDPhotonEmission.H>
```
```suggestion
    #include "Particles/ElementaryProcess/QEDInternals/QEDPairGeneration.H"
    #include "Particles/ElementaryProcess/QEDInternals/QEDPhotonEmission.H"
```

lucafedeli88 replied:
The path is actually just `Particles/ElementaryProcess/` , so I cannot commit directly your suggestion. However I have implemented the change.

lucafedeli88 commented on Source/Evolve/WarpXEvolve.cpp:
```diff
@@ -385,6 +385,11 @@ WarpX::OneStep_nosub (Real cur_time)
 
+#ifdef WARPX_QED
+    //Do QED processes
+    mypc->doQedEvents();
```
sure!

lucafedeli88 commented on Source/Particles/ElementaryProcess/QEDPairGeneration.H:
```diff
@@ -0,0 +1,157 @@
+    {
+        const auto opt_depth =
+            ptd.m_runtime_rdata[m_opt_depth_runtime_comp][i];
+        return (opt_depth < 0.0);
```
Sure! I've added `_rt` suffix also in other places in `QEDPairGeneration.H` and `QEDPhotonEmission.H`. `const auto` is now `const amrex::ParticleReal`, in order to be explicit...

lucafedeli88 commented on Source/Particles/ElementaryProcess/QEDPhotonEmission.H:
```diff
@@ -0,0 +1,203 @@
+
+        //Initialize the optical depth component of the source species.
+        src.m_runtime_rdata[m_opt_depth_runtime_comp][i_src] =
+            m_opt_depth_functor();
```
If I have understood correctly, `InitializationPolicy` specifies how the components of the product species should be initialized. However, here we need to reset the optical depth component of the source species. is there a way to use `InitializationPolicy`  also for source species?

MaxThevenet replied:
Oh no, sorry, you're absolutely right, that looks perfect.

lucafedeli88 commented:
I've also updated the documentation and QED-related examples

lucafedeli88 commented on Examples/Modules/qed/breit_wheeler/inputs_2d_tau_init:
```diff
@@ -62,6 +62,32 @@ photons.uz_th = 100.
+pos_bw.momentum_distribution_type = "gaussian"
+pos_bw.xmin = 1 ## Ugly trick to avoid positrons at T=0
+pos_bw.xmax = -1 ##  Ugly trick to avoid positrons at T=0
+pos_bw.do_qed = 0
```
I really like what you propose (including not allowing to create a species with just `photons.qed_breit_wheeler_ele_product_species = ele_bw`)!
Probably the easiest way to achieve this could be to allow for default values for most of the parameters of a species (except `mass` *and* `charge` or `species_type`).  The default value for `profile` could be something like `empty` or `do_not_generate`, so that the code could know that methods for particle initialization should not be called. 
Would something like that be OK? In case, I could work on a dedicated PR...

MaxThevenet replied:
Yes, that sounds good! You are most welcome to look into it.

lucafedeli88 commented:
Sorry, some tests are failing... I am trying to understand why...

lucafedeli88 commented:
Ok... issues have been fixed. There was an error related to 1/m , due to the fact that now m=0 for photons.

MaxThevenet commented on Source/Particles/MultiParticleContainer.cpp:
```diff
@@ -534,13 +535,57 @@ MultiParticleContainer::mapSpeciesProduct ()
+                allcontainers[i_product_phot]->AmIA<PhysicalSpecies::photon>(),
+                "ERROR: Quantum Synchrotron product species is of wrong type");
+        }
+#endif
```
Does this piece of code really belong here? Maybe we could put it in a new function
```c++
#ifdef WARPX_QED
MultiParticleContainer::CheckQEDSpecies(){
    ...
}
#endif
```
that we would call after `mapSpeciesProduct`, what do you think?

lucafedeli88 replied:
I have created a new `CheckQEDProductSpecies` function as suggested, which now includes all the checks (except the test on the existence of the product species). For consistency I have created an analogous function for the Ionization process. In the future we might decide to ask the user to mark as `electrons` the product species and check for that...

MaxThevenet commented on Source/Particles/MultiParticleContainer.cpp:
```diff
@@ -948,4 +974,137 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
+#endif
+            for (MFIter mfi = pc_source->MakeMFIter(lev, info); mfi.isValid(); ++mfi)
+            {
+                auto& src_tile = pc_source ->ParticlesAt(lev, mfi);
```
:) Thanks, sounds good.
```suggestion
                auto& src_tile = pc_source->ParticlesAt(lev, mfi);
```
^ Was this space here on purpose?

MaxThevenet commented on Source/Particles/MultiParticleContainer.cpp:
```diff
@@ -948,4 +974,137 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
+            }
+
+
+
```
```suggestion
```
Could we remove the three lines here? (I am not sure the Github suggestion is correct here, it might be trying to remove only one line, dunno why).

MaxThevenet commented on Source/Particles/MultiParticleContainer.cpp:
```diff
@@ -948,4 +974,137 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
+            static_cast<PhysicalParticleContainer*>(pc_source.get());
+
+        const auto Filter    = phys_pc_ptr->getPhotonEmissionFilterFunc();
+        const auto CopyPhot   = copy_factory_phot.getSmartCopy();
```
```suggestion
        const auto Filter   = phys_pc_ptr->getPhotonEmissionFilterFunc();
        const auto CopyPhot = copy_factory_phot.getSmartCopy();
```

MaxThevenet commented on Source/Particles/MultiParticleContainer.cpp:
```diff
@@ -948,4 +974,137 @@ MultiParticleContainer::BreitWheelerGenerateTable ()
+                setNewParticleIDs(dst_tile, np_dst, num_added);
+
+                constexpr auto energy_threshold = static_cast<amrex::ParticleReal>(
+                     2.0 * PhysConst::m_e * PhysConst::c * PhysConst::c);
```
Should this be an input parameter?

lucafedeli88 replied:
yes, we could make it an input parameter... I will do that

lucafedeli88 replied:
I should have implemented the change. The energy threshold for photon creation is now an
optional input parameter. I have updated the documentation accordingly.

MaxThevenet commented on Source/Particles/ElementaryProcess/QEDPhotonEmission.H:
```diff
@@ -0,0 +1,208 @@
+void cleanLowEnergyPhotons(
+PTile& ptile,
+const int old_size, const int num_added,
+const amrex::ParticleReal t_energy_threshold)
```
```suggestion
    PTile& ptile,
    const int old_size, const int num_added,
    const amrex::ParticleReal t_energy_threshold)
```
or am I missing some indentation conventions?

lucafedeli88 replied:
oh sorry! You are right!

MaxThevenet commented on Source/Particles/ElementaryProcess/QEDPhotonEmission.H:
```diff
@@ -0,0 +1,208 @@
+
+    const auto energy_threshold =
+        t_energy_threshold/(PhysConst::m_e*PhysConst::c);
+    const auto energy_threshold2 = energy_threshold*energy_threshold;
```
Could you comment on what you are doing with the physical quantities here? `t_energy_threshold` is a physical energy, correct? So `energy_threshold` is a velocity? Maybe some comments in the code would help :).

lucafedeli88 replied:
I have re-organized the calculation, so that it should appear more transparent now...

MaxThevenet commented on Source/Particles/MultiParticleContainer.cpp:
```diff
@@ -562,10 +583,14 @@ MultiParticleContainer::getSpeciesID (std::string product_str)
     );
 
+>>>>>>> upstream/master
     return i_product;
```
Some merge markers have survived here.

MaxThevenet commented:
:clap:
